### PR TITLE
Warn about `for...in`->`for...Object.entries`-style mismatch & null guards

### PR DESCRIPTION
**Pull request by GoodForOneFare**
Adds some really verbose warnings about for loop conversions.  I think the wording needs a lot of work.

Sample output:
shopify zsh 2016-11-24 11-16-27

Please review: @lemonmade, @TylerHorth

### CODE REVIEW

File: packages/esify/lib/decaf.js [hunk 1]
```diff
@@ -39,11 +39,31 @@ var WARNING_CHECKS = [
...
+    var loopMatcher = /for \(.+ of Object\.(entries|keys)/g;
+    var methods = new Set();
+    var match = loopMatcher.exec(source);
+    while (match != null) {
+      methods.add(match[1]);
+      match = loopMatcher.exec(source);
+    }
+
+    methods.forEach((method) => {
+      warn('Your file contains a `for` loop that has been translated to loop over `Object.' + method + '`. Previously, CoffeeScript would translate this to a `for...in` loop, which loops through non-iterable properties.  If the object being passed into `entries` is not an array/map, please convert the loop back to a `for...in` loop.');
```
Comment (TylerHorth) [inline packages/esify/lib/decaf.js]:
```diff
-...If the object being passed into `entries` is not an array/map...
+...If the object being passed into `' + method +'` is not an array/map...
```

File: packages/esify/lib/decaf.js [hunk 2]
```diff
@@ -39,11 +39,31 @@ var WARNING_CHECKS = [
   },
 ];
 
+const POST_DECAF_CHECKS = [
+  function checkForMistranslatedForLoops(source) {
+    var loopMatcher = /for \(.+ of Object\.(entries|keys)\([^)]+ \|\| \{\}\)/g;
```
Comment (TylerHorth) [inline packages/esify/lib/decaf.js]: You don't need to escape `{` or `}`.

File: packages/esify/lib/decaf.js [hunk 3]
```diff
@@ -39,11 +39,31 @@ var WARNING_CHECKS = [
...
+    var methods = new Set();
+    var match = loopMatcher.exec(source);
+    while (match != null) {
+      methods.add(match[1]);
+      match = loopMatcher.exec(source);
+    }
+
+    methods.forEach((method) => {
+      warn('Your file contains a `for` loop that has been translated to loop over `Object.' + method + '`. Previously, CoffeeScript would translate this to a `for...in` loop, which loops through non-iterable properties.  If the object being passed into `entries` is not an array/map, please convert the loop back to a `for...in` loop.');
+      warn('A fallback `|| {}` value has been added to your `Object.' + method + '` call. If the values being passed into `' + method + '` is guaranteed to be non-null, it is safe to remove the fallback.');
```
Comment (TylerHorth) [inline packages/esify/lib/decaf.js]: `s/the values being passed/the value being passed/`

**Comment by TylerHorth** [comment]
I think It's alright to be verbose in this instance. Someone not familiar with modern javascript is going to have a hard time figuring out what they're supposed to do otherwise.

**Comment by lemonmade** [comment]
As Tyler said, I'm fine with verbosity given the complexity of the things you are warning about.

**Pull request merged**